### PR TITLE
Add a callback for the final failure

### PR DESCRIPTION
**Pull request by jcobian**
Solves #72

### CODE REVIEW

**Comment by jcobian** [comment]
Travis is currently failing b/c of a rubygems issue but that should be fixed once https://github.com/kamui/retriable/pull/75 is merged

**Comment by jcobian** [comment]
@kamui thoughts about this change/interface?

File: README.md [hunk 1]
```diff
@@ -202,14 +203,18 @@ end
 
 ### Callbacks
 
-`#retriable` also provides a callback called `:on_retry` that will run after an exception is rescued. This callback provides the `exception` that was raised in the current try, the `try_number`, the `elapsed_time` for all tries so far, and the time in seconds of the `next_interval`. As these are specified in a `Proc`, unnecessary variables can be left out of the parameter list.
+`#retriable` also provides callbacks called `:on_retry`  and `:on_final_failure` that will run after an exception is rescued.  The `:on_retry` callback provides the `exception` that was raised in the current try, the `try_number`, the `elapsed_time` for all tries so far, and the time in seconds of the `next_interval`. The `:on_final_failure` callback provides everything but the `next_interval`. As these are specified in a `Proc`, unnecessary variables can be left out of the parameter list.
```
Comment (apurvis) [inline README.md]: "after an exception is rescued or fails, respectively"

File: spec/retriable_spec.rb [hunk 2]
```diff
@@ -222,6 +222,22 @@ def increment_tries_with_exception(exception_class = nil)
...
+    context "with an on_final_failure handler" do
+      it "calls the on_final_failure proc once at the end" do
+        exceptions = []
+        handler = ->(exception, try, _elapsed_time) { exceptions[try] = exception }
+        expect do
+          described_class.retriable(on_final_failure: handler, tries: 3) do
+            increment_tries
+            raise StandardError, "foo"
+          end
+        end.to raise_error(StandardError, /foo/)
```
Comment (apurvis) [inline spec/retriable_spec.rb]: maybe put some whitespace after this line?

File: README.md (hunk 1 above)
Comment (jcobian) [reply README.md]: 👍 . Updated but let me know if you were thinking of less than what I have now as well. Just want to make it clear that it doesn't pass `next_interval` to the `on_final_failure` proc as well

**Comment by kamui** [comment]
Thinking about this a bit, there could be 2 issues:

1. From the naming and interface, it's not clear without reading the code or adding documentation that  on the final error runs before/after the final `on_retry` call. Both are hooks that run after rescue.
2. If you perhaps want to do something different on the last failure, but perhaps you also want the `on_retry` callback to also act differently too. I'm not sure this is a real issue, but I've not needed to a final failure callback before, so I don't have a good grasp on how people might use this.

I'd like to throw out an idea that could solve both issues, but might be less neat. What if we just passed a boolean block param (`is_final_failure`) into `on_retry` callback instead of adding another callback? Or the param could be `tries_remaining` which might be useful for other things too. Then you could alter the last `on_retry` callback by code branching on that parameter, you can be entirely flexible on the final callback execution and there's no confusion when it gets run because it's entirely in the user's control.

Thoughts?

**Comment by apurvis** [comment]
> i like the idea of passing `tries_remaining`

just reiterating; i think it's definitely a cleaner solution with a really nice side effect that i have wanted before when using the gem.

**Comment by jcobian** [comment]
hi -- sorry i didn't get around to this and don't have the time to do this. Going to close this out.

**Comment by timolehto** [comment]
Hey, sorry to resurrect this 🧟. Maybe I should just create a new issue or something, but since we have all the context here...

@apurvis and @kamui how would/should `tries_remaining` work in the case where `max_elapsed_time` runs out? It just jumps from some number such as 35 down to 0?

Also we already have the `next_interval` value passed to the callback. I would have expected that to be `nil` in the case where we have run out of time (like it is when we run out of tries), but it's not. If it was then it would already provide us with the means to detect the last error case, though it's maybe not the most intuitive thing.. In any case I wonder if this is a bug and we could/should just first check the end condition here
```
        on_retry.call(exception, try, elapsed_time.call, interval) if on_retry
        raise if try >= tries || (elapsed_time.call + interval) > max_elapsed_time
```
https://github.com/kamui/retriable/blob/c6ccd44f8ac2e2b7e44f1d8f812305caa7a2f5cd/lib/retriable.rb#L70:L71
and if it's time to end then set interval to `nil` before passing to `on_retry` and then raise after?

What do you think?

**Pull request closed without merging**